doubly_linked_list/doubly_linked_list.py

```diff
@@ -1,68 +1,163 @@
+    """Wrap the given value in a ListNode and insert it
   after this node. Note that this node could already
   have a next node it is point to."""
-  def insert_after(self, value):
-    current_next = self.next
-    self.next = ListNode(value, self, current_next)
-    if current_next:
-      current_next.prev = self.next
 
-  """Wrap the given value in a ListNode and insert it
+    def insert_after(self, value):
+        current_next = self.next
+        self.next = ListNode(value, self, current_next)
+        if current_next:
+            current_next.prev = self.next
+
+    """Wrap the given value in a ListNode and insert it
   before this node. Note that this node could already
   have a previous node it is point to."""
-  def insert_before(self, value):
-    current_prev = self.prev
-    self.prev = ListNode(value, current_prev, self)
-    if current_prev:
-      current_prev.next = self.prev
 
-  """Rearranges this ListNode's previous and next pointers
+    def insert_before(self, value):
+        current_prev = self.prev
+        self.prev = ListNode(value, current_prev, self)
+        if current_prev:
+            current_prev.next = self.prev
+
+    """Rearranges this ListNode's previous and next pointers
   accordingly, effectively deleting this ListNode."""
-  def delete(self):
-    if self.prev:
-      self.prev.next = self.next
-    if self.next:
-      self.next.prev = self.prev
+
+    def delete(self):
+        if self.prev:
+            self.prev.next = self.next
+        if self.next:
+            self.next.prev = self.prev
+
 
 """Our doubly-linked list class. It holds references to
 the list's head and tail nodes."""
+
+
 class DoublyLinkedList:
-  def __init__(self, node=None):
-    self.head = node
-    self.tail = node
-    self.length = 1 if node is not None else 0
+    def __init__(self, node=None):
+        self.head = node
+        self.tail = node
+        self.length = 1 if node is not None else 0
+
+    def __len__(self):
+        return self.length
+
+    def add_to_head(self, value):
+        current_head = self.head
+        new_head = ListNode(value)
+
+        if not current_head:
+            self.head = new_head
+        else:
+            current_head.insert_before(value)
+            self.head = current_head.prev
+            self.length += 1
+
+    def remove_from_head(self):
+
+        if not self.head and not self.tail:
+            return None
+
+        if self.head.next is None:
+            removed_head = self.head.value
+            self.head = None
+            self.tail = None
+            self.length -= 1
+            return removed_head
+
+        else:
+            removed_head = self.head.value
+            self.head = self.head.next
+            self.head.prev = None
+            self.length -= 1
+
+            return removed_head
 
-  def __len__(self):
-    return self.length
+    def add_to_tail(self, value):
+        if self.tail is None:
```

Comment:
good job writing out the edge cases first, and then progressing down for the typical case!